src/main/java/ru/testtask/service/FileService.java

```diff
@@ -0,0 +1,250 @@
+        }
+
+    public List<FileData> searchByRegex(String regex){
+        return fileDataRepo.findFileDataByRegexpFilename(regex);
+    }
+
+    public FileData moveFile(String id, String directory) throws FileNotFoundException {
+        Optional<FileData> optionalFileData = fileDataRepo.findById(id);
+        if (optionalFileData.isEmpty()){
+            throw new FileNotFoundException(String.format("File with id %s does not found", id));
+        }
+            FileData fileData = optionalFileData.get();
+            Path destDirectory = storageRootPath.resolve(directory);
+            Path destination = destDirectory.resolve(fileData.getFilename());
+
+        try {
+            if (!Files.exists(destDirectory)) {
+                Files.createDirectories(destDirectory);
+            }
+
+            Files.move(getFileAbsolutePath(fileData), destination);
+            fileData.setDirectory("/" + directory);
+            fileDataRepo.save(fileData);
+        } catch (IOException e) {
+            log.error(String.format("An error occurred during moving the file with id %s", id));
+        } catch (MongoWriteException e){
+            log.error("File with the same directory and filename already exists!");
+        }
+
+        return fileData;
+    }
+
+    public void copyFile(String id, String directory) throws IOException, FileNotFoundException {
+        String newOriginalFileName;
+        int i = 0;
+        Optional<FileData> optionalFileData = fileDataRepo.findById(id);
+
+        if (optionalFileData.isEmpty()){
+            throw new FileNotFoundException(String.format("File with ID %s does not found", id));
+        }
+
+        Path destDirectory = Paths.get(storageRoot + File.separator + directory);
+
+        FileData fileData = optionalFileData.get();
+        String newFilename = UUID.randomUUID().toString() + "." + FilenameUtils.getExtension(fileData.getFilename());
+        Path destination = destDirectory.resolve(fileData.getFilename());
+
+        try {
+            if (!Files.exists(destDirectory)) {
+                Files.createDirectories(destDirectory);
+            }
+
+            Files.copy(getFileAbsolutePath(fileData), destination);
+        } catch (IOException e) {
+            log.error(String.format("An error occurred during copying the file with id %s", id));
+        }
+
+        List<FileData> fileDatas = fileDataRepo.findFileDataByRegexpFilename(FilenameUtils
+                .removeExtension(fileData.getOriginalFilename()));
+        List<String> fileDataListNames = fileDatas.stream().map(FileData::getOriginalFilename).collect(Collectors.toList());
+
+        do {
+              newOriginalFileName = FilenameUtils.removeExtension(fileData.getOriginalFilename())
+                      + " (" + i + ")." + FilenameUtils.getExtension(fileData.getOriginalFilename());
+            i++;
+        }
+        while(fileDataListNames.contains(newOriginalFileName));
+
+        fileData.setFilename(newFilename);
```

Comment:
What happens if two threads will calculate similar original file name, and try to save calculated data?